### DISCUSSION TITLE
Import GPX with waypoints, route, track selection

### DISCUSSION THREAD

**Discussion by MotoUKRider** [devemux86/cruiser#44] [Q&A] [answered]
*(continued from another topic)*

Purely an observation/discussion rather than a "needed" feature, but this is what *can* be exported from Garmin's BaseCamp. I don't know how many people do it this way, but I found out the following purely by chance...

Garmin's BaseCamp allows you to select "things" to export, usually as a GPX.  Typically, this is a Route of a ride/trip somewhere.  I also use it to filter all my Waypoints (stops, photo points etc) into a single GPX which is then imported into Cruiser as "Favorites" - works brilliantly.  If you select multiple items during export, the GPX file gets quite complicated.... so here goes:

Opening a GPX file into ITNConv that contains multiple routes will offer a content list of what is inside the file: So, a single GPX file could be a full 7-day tour with 7 routes inside it plus the list of all the Via/Waypoints.

In the attached example: 3 routes and a set of Waypoints (eg: favourites/locations) from a multi-day tour of lower Scotland.  In BaseCamp, I selected the 3 Routes and exported them. The resulting GPX is huuuuuge, but offered for analysis.

[multiple-routes.zip](https://github.com/devemux86/cruiser/files/10776564/multiple-routes.zip)

Screen from ITNConv showing what it can "see":
![gpx-multiple-routes-1](https://user-images.githubusercontent.com/125696109/219942202-cf1a4980-80cd-435c-8e68-0dc0f3d37667.jpg)

Loaded "Day2" from that list:
![gpx-multiple-routes-2](https://user-images.githubusercontent.com/125696109/219942208-97428f51-ec7d-4b9a-a4d1-5c8db3a4bb62.jpg)

PS: In BaseCamp, you can't have duplicate names of anything, so you'll see in the routing list that B7042+B70421 and A719+A7192 are actually the same road but have a number suffix. I obviously have "A7191" in another route somewhere. This can be annoying if you want to put a custom text tag on a shaping point that you might want to switch to a Via Point, eg: "Turn off here" - so you may find that road names/numbers often have a number at the end if they've come from BaseCamp. Fun!

**Comment by devemux86** [comment]
Thanks for the suggestion!

Indeed, this seems like a useful feature to better filter what to import from gpx files that have many routes or tracks.

I have put it in my list to see how we can enrich the import process.

**Comment by 0709wiwiwi** [comment]
**A Cruiser forum, so think Cruiser, design with Cruiser, and navigate with Cruiser**  

Multi navigation track support this in a single very **simple** to understand gpx file.

What is already possible by the simple to operatee but powerfull Cruiser app ?  
What eventually still can be improved  ? That's the question. So lets have a look.
(Any nice ideas are welcome)

- Multi navigation tracks support is definitely quite possible with the Cruiser app.
- Multi navigation tracks with all the necessary present this in a single gpx file. 
- This can even include individual so freely located poi waypoints without fail.  

A multi gpx navigation track zipped is even quite compact in file transfers.

Multi tracks contain and Turnpoints and Shaping points and the Via points. 
All this nicely tightly organised per individual navigation track. 
All Shaping or Via points can then even comfortably be identically named. 
As these planner (routing- points are directly connected to the track point, so their position will never be incorrect.

Multitrack gpx support is thus simple this without external generation nor any conversion required. 
This way of simple construction has not yet been proposed as a new idea in this Cruiser forum. 
As Rome too, by the way, was not built in a day either. So some patience does pay off.

An example of a multitrack gpx is attached. Alternative navigation tracks with same start and end. 
Multi navigation tracks that continue this as a single route (day trips) can be done just as well.

Moreover, freely located added poi wpt can very easily be added in the same gpx file.
These are currently now shown as blue dots in the Cruiser map view.  (No specific Icon support)
These waypoints simple belong perfectly to the current imported multitracks gpx file. 

Free waypoints with specific Icon support (future) that are specified as poi or favourites points. 
Such Icons support per category can later still be further optimised into the Cruiser app.
However, this again concerns a completely different item new idea proposal to be supported.

Willy.
[Ediger_Eller to Cochem navtrack A_B_C_D and a single demo poi wpt.gpx.txt](https://github.com/devemux86/cruiser/files/10776994/Ediger_Eller.to.Cochem.navtrack.A_B_C_D.and.a.single.demo.poi.wpt.gpx.txt)


Translated with www.DeepL.com/Translator (free version)

**Reply by devemux86** [reply]
> Multi navigation tracks support is definitely quite possible with the Cruiser app

Multiple routes in **navigation gpx** are already supported!

Cruiser can import (offline) this navigation gpx and all 4 routes are available at once:

<img src="https://user-images.githubusercontent.com/3484020/219952752-6111f2d8-c49a-41ba-a80b-d1483ba24531.jpg" width="50%"/><img src="https://user-images.githubusercontent.com/3484020/219952756-21a9e95b-d1a0-4db3-994b-27cb0f2e8fb9.jpg" width="50%"/>

**Comment by 0709wiwiwi** [comment]
I know that. The reply was mainly intended to point out to the questioner what Cruiser can also do. 

So think a little more Cruiser, and also design especially using the Cruiser app and possibly think a little (a little)  less so "Garmin". 
This concerns the support forum for the Cruiser app and the further independent development of this app. 
I don't have to see a Garmin copy here. It is precisely that complexity that I want to avoid with the Cruiser app. 
And please don't get me wrong,  (I did use Garmin gear before excellent hardware by the way)
So I'm not saying a bad word about that very big and very important commercial player, nor about the used methods but I mainly support this very promising alternative, especially here.  A new alternative ! Give it a try and support so Get Pro versions and this for 10 euro only.

**Reply by MotoUKRider** [reply]
It's worth remembering that GPX is a very popular format (if not the standard), and if the formatted file is *"within spec"*, then supporting it should at least be considered. Other applications not from Garmin could also make these multi-route/track files, hence my opening statement on the original message :)

**Reply by 0709wiwiwi** [reply]
Know not everything you find on the internet is conform with the gpx 1.1 standard.
When the strict Garmin programs ACCEPT imports, the file is mostly "_within spec_"

**Comment by 0709wiwiwi** [comment]
Indeed, multitrack gpx import in Cruiser is already very practical and definitely nicely presented. 
To be able to transfer multiple navigation tracks this in a single gpx file to share sure is a nice to have.

"The gpx file demo example shown I produced this one by using an external program.

Merging multiple navtracks for export into a single gpx file together with individual selected poi wpts has not yet been suggested so far. 
So here it is.  **NEW Cruiser suggestion** ;-) 

Route export > Set Navigation:

![image](https://user-images.githubusercontent.com/54167201/220056835-55124153-7157-42ae-8b38-0f220aef1f6a.png)

Above all, please thereby also keep that gpx (poi) waypoint list very clean and most compact. 
Without unnecessary duplicates (blue dots) of both the turn and planner (shaping_via) trkpts.

Regards, 
Willy.


Translated with www.DeepL.com/Translator (free version)

**Reply by devemux86** [reply]
> Merging multiple navtracks into a single gpx file together with individual selected poi wpts has not yet been suggested so far.

Thanks for the suggestion!

Can you describe the source of these poi?
Do you mean favorites?

@0709wiwiwi 
As this discussion is about "Import GPX" and you describe "Export GPX" (?)
better to create a new topic so we can discuss it there?

**Comment by 0709wiwiwi** [comment]


**Reply by devemux86** [reply]
@0709wiwiwi 
I don't think GitHub offers moving posts (?).

Please create a new topic with the description of the suggestion. :slightly_smiling_face:

**Comment by 0709wiwiwi** [comment]
A lot starts with imagination so here is a contribution of that kind. 
Usefully to be used or maybe not ?  Please read the document and comment.

 Video: [Cruiser navigate merged navtracks](https://youtu.be/x4H3OOWTj0E)

[Cruiser merged gpx  navtrack navigation fantasy.docx](https://github.com/devemux86/cruiser/files/12196938/Cruiser.merged.gpx.navtrack.navigation.fantasy.docx)

[Merged_large_gap_navtracks.gpx.txt](https://github.com/devemux86/cruiser/files/12196989/Merged_large_gap_navtracks.gpx.txt)

**Comment by devemux86** [comment]
**Import** was rewritten, the new import is coming in next Cruiser:

- Select directly the file to import (format is auto detected)

- Formats: gpx, gpz, geojson, json, kurviger, itn

- Import **analyzes** the GPX file and lists its items with names:
  - Navigation tracks
  - Tracks
  - Routes
  - Waypoints

- Select a GPX item to import for routing

For example this complex [test.gpx.txt](https://github.com/user-attachments/files/17502210/test.gpx.txt) with navigation and normal tracks, routes and waypoints:

<img src="https://github.com/user-attachments/assets/13db591b-4cd2-4f31-aef7-9a1b04405c98" width="50%"/><img src="https://github.com/user-attachments/assets/58ee8e3e-f6f3-4215-8842-5d273f54faeb" width="50%"/>

**Reply by bios4** [reply]
This sounds great! 👍

> Select directly the file to import (format is auto detected)

In this case when it could be gpx/gpz, json, kurviger, itn and the like I would prefer NOT to read "GPX" in the dialog window, but something more generic. Maybe "Source file" (or just "Source") or something like this!?

<img src="https://github.com/user-attachments/assets/5b7b45d0-e762-43d2-9db8-5d59804d1df5" width="33%">

**Reply by devemux86** [reply]
The selection only exists when import GPX (or zipped gpx as gpz),
as only the GPX format may store multiple data types to select.

All other formats provide a single data type to process:
- GeoJSON: route imported without calculation from BRouter website
- JSON / Kurviger: route imported without calculation
- ITN: waypoints to calculate route

**Reply by bios4** [reply]
So your screenshots above show the dialog AFTER a GPX (GPZ) file was already selected!?

How does the UI then look like right after initiating an import (before Cruiser knows that a GPX should be processed)?

Or does the new import open just a default Andoid file selection dialog?

**Reply by devemux86** [reply]
- The 1st step of the import is now the Android file browser to select a file

- Cruiser analyzes the format and presents the import dialog with options:
  - GPX / GPZ: there are options to select the data type and max waypoints
  - Other formats: only the overlay / routing option

**Reply by bios4** [reply]
> The 1st step of the import is now the Android file browser to select a file

When initiating an import (either by Routing > Import or by long-press navigation button > Import) the following screen is opened BEFORE any file selection browser, asking for a file type (and then only showing files of the selected type in the next file browser window).

So what am I doing wrong @devemux86?
How is that auto-detect feature supposed to work?
Do I have to adjust a setting somewhere?

5.0.1-b12 by the way...

<img src="https://github.com/user-attachments/assets/0eafa866-8df4-4ae9-9e7f-f8ef8b60a58d" width="33%">

**Reply by devemux86** [reply]
@bios4 This feature was not announced in any release, it will probably be included in Cruiser **5.1**.

**Reply by bios4** [reply]
Ah okay, I thought that this is already in testing phase with the latest beta(s)...

Thanks for clarifying!

**Comment by devemux86** [comment, marked as answer]
- Import GPX with waypoints, route, track selection is available in https://github.com/devemux86/cruiser/discussions/684
